packages/wallet/test/services/WalletUtil.test.ts

```diff
@@ -607,5 +613,344 @@ describe('WalletUtil', () => {
+            }
+          } as Cardano.StakeRegistrationDelegationCertificate
+        ];
+
+        expect(await requiresForeignSignatures(tx, scriptWallet)).toBeTruthy();
+      });
+
+      it('detects foreign VoteRegistrationDelegation', async () => {
+        tx.body.certificates = [
+          {
+            __typename: Cardano.CertificateType.VoteRegistrationDelegation,
+            stakeCredential: {
+              hash: Crypto.Hash28ByteBase16.fromEd25519KeyHashHex(foreignRewardAccountHash),
+              type: Cardano.CredentialType.KeyHash
+            }
+          } as Cardano.VoteRegistrationDelegationCertificate
+        ];
+
+        expect(await requiresForeignSignatures(tx, scriptWallet)).toBeTruthy();
+      });
+
+      it('detects foreign stake_vote_reg_deleg_cert', async () => {
+        tx.body.certificates = [
+          {
+            __typename: Cardano.CertificateType.StakeVoteRegistrationDelegation,
+            stakeCredential: {
+              hash: Crypto.Hash28ByteBase16.fromEd25519KeyHashHex(foreignRewardAccountHash),
+              type: Cardano.CredentialType.KeyHash
+            }
+          } as Cardano.StakeVoteRegistrationDelegationCertificate
+        ];
+
+        expect(await requiresForeignSignatures(tx, scriptWallet)).toBeTruthy();
+      });
+
+      it('detects foreign Unregistration', async () => {
+        tx.body.certificates = [
+          {
+            __typename: Cardano.CertificateType.Unregistration,
+            stakeCredential: {
+              hash: Crypto.Hash28ByteBase16.fromEd25519KeyHashHex(foreignRewardAccountHash),
+              type: Cardano.CredentialType.KeyHash
+            }
+          } as Cardano.NewStakeAddressCertificate
+        ];
+
+        expect(await requiresForeignSignatures(tx, scriptWallet)).toBeTruthy();
+      });
+
+      it('returns true when at least one input from collateral is not accounted for ', async () => {
+        tx.body.collaterals = [
+          {
+            index: 0,
+            txId: Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000')
+          },
+          ...tx.body.collaterals!
+        ];
+
+        expect(tx.body.inputs.length).toBeGreaterThanOrEqual(1);
+        expect(tx.body.certificates!.length).toBe(1);
+        expect(await requiresForeignSignatures(tx, scriptWallet)).toBeTruthy();
+      });
+
+      it('returns true when at least one input is not accounted for ', async () => {
+        tx.body.inputs = [
+          {
+            index: 0,
+            txId: Cardano.TransactionId('0000000000000000000000000000000000000000000000000000000000000000')
+          },
+          ...tx.body.inputs
+        ];
+
+        expect(tx.body.inputs.length).toBeGreaterThanOrEqual(2);
+        expect(tx.body.certificates!.length).toBe(1);
+        expect(await requiresForeignSignatures(tx, scriptWallet)).toBeTruthy();
+      });
+
+      it('Voting procedures are always foreign', async () => {
```

Comment:
Maybe we could implement the proper check that compares script hash for script credentials on this level. It won't be possible to use via cip30, but that would be a more correct implementation, right? Same with DRepCredential and probably some voting certificate types

Reply:
done here https://github.com/input-output-hk/cardano-js-sdk/pull/1030/commits/508d7e4f778e8ac6f98c0905a8e7dca178e49b56